### PR TITLE
added breaking changes for the Java API to the breaking changes doc for 5.0

### LINKED ISSUE #14191: Document breaking changes introduced with #13752
With #13752 we introduced quite some breaking changes to the java api. We should review what we did and properly document those breaking changes in the migration guide for 3.0. Also, we should consider deprecating in 2.x some the methods that we have removed from `SearchRequest` and `SearchRequestBuilder` etc.

### PR DESCRIPTION
**Pull request by colings86**
Closes #14191

### CODE REVIEW

File: docs/reference/migration/migrate_5_0/java.asciidoc [hunk 1]
```diff
@@ -225,3 +225,40 @@ The `addSuggestion` method now required the user specified suggestion name, prev
 ===== SuggestionBuilder
 
 The `field` setter has been deleted. Instead the field name needs to be specified as constructor argument.
+
+==== SearchSourceBuilder
+
+All methods which take an `XContentBuilder`, `BytesReference` `Map<String, Object>` or `bytes[]` have been removed in favor of providing the
+relevant builder object for that feature. This means that all search requests can now be validated at call time which results in much clearer
```
Comment (javanna) [inline docs/reference/migration/migrate_5_0/java.asciidoc]: can you add an example of what we mean by builder object? Also do we still want to remove the builder suffix? In that case we may want to start here by using a different name?

**Comment by javanna** [comment]
left a small comment, thanks a lot for doing thins Colin

**Comment by javanna** [comment]
left a small comment, thanks a lot for doing thins Colin

**Comment by javanna** [comment]
One more thing, do you mind adding the same things for validate query and explain?
See https://github.com/elastic/elasticsearch/issues/14191#issuecomment-152714057

**Comment by colings86** [comment]
@javanna I pushed a commit to address your comments

File: docs/reference/migration/migrate_5_0/java.asciidoc [hunk 2]
```diff
@@ -229,16 +229,16 @@ The `field` setter has been deleted. Instead the field name needs to be specifie
 ==== SearchSourceBuilder
 
 All methods which take an `XContentBuilder`, `BytesReference` `Map<String, Object>` or `bytes[]` have been removed in favor of providing the
-relevant builder object for that feature. This means that all search requests can now be validated at call time which results in much clearer
-errors.
+relevant builder object for that feature (e.g. HighlightBuilder, AggregationBuilder, SuggesterBuilder) . This means that all search requests
```
Comment (javanna) [inline docs/reference/migration/migrate_5_0/java.asciidoc]: is it SuggesterBuilder or SuggestBuilder?

File: docs/reference/migration/migrate_5_0/java.asciidoc [hunk 3]
```diff
@@ -225,3 +225,58 @@ The `addSuggestion` method now required the user specified suggestion name, prev
...
 
 The `field` setter has been deleted. Instead the field name needs to be specified as constructor argument.
+
+==== SearchSourceBuilder
+
+All methods which take an `XContentBuilder`, `BytesReference` `Map<String, Object>` or `bytes[]` have been removed in favor of providing the
+relevant builder object for that feature (e.g. HighlightBuilder, AggregationBuilder, SuggesterBuilder) . This means that all search requests
+can now be validated at call time which results in much clearer errors.
+
+The `defaultResourceWindowSize(int)` method has been removed. The window size should be set explicitly on all RescoreBuilder objects.
```
Comment (javanna) [inline docs/reference/migration/migrate_5_0/java.asciidoc]: you may want to use the backtick for RescoreBuillder, same for HighlightBuilder, AggregationBuilder and SuggestBuilder wherever they are

File: docs/reference/migration/migrate_5_0/java.asciidoc [hunk 4]
```diff
@@ -225,3 +225,58 @@ The `addSuggestion` method now required the user specified suggestion name, prev
...
+`setRescorer(Rescorer)` and `setRescorer(Rescorer, int)` have been removed infavor of `setRescorer(RescoreBuilder)` and `setRescorer(RescoreBuilder, int)`
+
+==== SearchRequest
+
+All template methods have been removed in favor of a single `template(Template)` method.
+
+All source methods have been removed in favor of a single `source(SearchSourceBuilder)` method. This means that all search requests can now be validated
+at call time which results in much clearer errors.
+
+All extraSource methods have been removed.
```
Comment (javanna) [inline docs/reference/migration/migrate_5_0/java.asciidoc]: backtick here as well?

**Comment by javanna** [comment]
left very small comments, really nitpicks, LGTM though no need for another review

**Comment by javanna** [comment]
Do we want to go and deprecate these methods in the last 2.x release? That would probably be nice to java api users, as their code will not compile anymore against 5.0.

**Pull request merged**